Library/Homebrew/livecheck/strategy/hackage.rb

```diff
@@ -17,8 +17,19 @@ module Strategy
       #
       # @api public
       class Hackage
-        # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = /(?:downloads|hackage)\.haskell\.org/i.freeze
+        # A `Regexp` used in determining if the strategy applies to the URL and
+        # also as part of extracting the package name from the URL basename.
+        PACKAGE_NAME_REGEX = /(?<package_name>.+?)-\d+/i.freeze
+
+        # A `Regexp` used to extract the package name from the URL basename.
+        FILENAME_REGEX = /^#{PACKAGE_NAME_REGEX.source.strip}/i.freeze
+
+        # A `Regexp` used in determining if the strategy applies to the URL.
+        URL_MATCH_REGEX = %r{
+          ^https?://(?:downloads|hackage)\.haskell\.org
+          /.+
```

Comment:
```rb
# Hackage URLs take one of the following formats:
#
# * `https://hackage.haskell.org/package/example-1.2.3/example-1.2.3.tar.gz`
# * `https://downloads.haskell.org/~ghc/8.10.1/ghc-8.10.1-src.tar.xz`
```
There are always exactly two path segments before the filename:
```suggestion
          (?:/[^/]+){2}
```

Reply:
Same as above. We're interested in the filename at the end of the path, so the number of directories doesn't matter all that much. The existing URL examples use two directories but this isn't necessarily guaranteed, so I think there's something to be said for being flexible in parts of the regex that aren't as important. I could understand if we were matching more URLs than we should be but that isn't the case yet.

In general, it's arguably easier to identify when a strategy is being applied to a URL it shouldn't match compared to identifying when a strategy isn't being applied to a URL it should match. Detecting the presence of something is often easier in livecheck than trying to detect the absence, so that's also part of my reasoning in terms of erring on the side of looseness here.

Reply:
Yeah, we are successfully bikeshedding about regexes ATM but I find it truly useful within a good measure and with the right people.

> Detecting the presence of something is often easier in livecheck than trying to detect the absence, so that's also part of my reasoning in terms of erring on the side of looseness here.

That is true but it also means detecting false presence is easier too same as detecting false absence. My today's _hot take_ is that regexes aren't really about detecting presence/absence but matching (possibly fully qualified) form. It is borderline comparable to designing a keyhole pattern from an existing key, and it is always a good idea to be specific as possible to ensure only your (set of) key/s opens that lock. Except obviously, livecheck doesn't protect anyone's money safe but that's how I formed my opinion :upside_down_face:

Reply:
So in the spirit of consistency and with acknowledging your arguments, here comes another proposal:
```suggestion
          (?:/[^/]+)+
```

This is a change for sake of change but I believe (yes it is subjective :shrug:) it better conveys the intent to the fellow programmer.

Reply:
Actually, following your line of thinking it would probably be better to invert it by changing the PyPI one to `/.+`. Whatever, but let's make it consistent if possible.

Reply:
They both match the same text in this context but `(?:/[^/]+)+` is clear about what type of text it's working with, so I'm good with the replacement in `Hackage`. I incorporated it in the latest push.